src/test/java/org/broadinstitute/hellbender/engine/ReadsDataSourceUnitTest.java

```diff
@@ -532,4 +538,54 @@ private File getFileWithAddedContig(
         return outputFile;
     }
 
+    @DataProvider(name = "manuallySpecifiedIndexTestData")
+    public Object[][] manuallySpecifiedIndexTestData() {
+        final String BAM_DIR = READS_DATA_SOURCE_TEST_DIRECTORY + "readIndexTest/";
+        final String INDEX_DIR = BAM_DIR + "indices/";
+
+        final List<Path> bams = Arrays.asList(IOUtils.getPath(BAM_DIR + "reads_data_source_test1.bam"),
+                IOUtils.getPath(BAM_DIR + "reads_data_source_test2.bam"));
+
+        final List<Path> indices = Arrays.asList(IOUtils.getPath(INDEX_DIR + "reads_data_source_test1.bam.bai"),
+                IOUtils.getPath(INDEX_DIR + "reads_data_source_test2.bam.bai"));
+
+        return new Object[][]{
+                {bams, indices}
+        };
+    }
+
+    @Test(dataProvider = "manuallySpecifiedIndexTestData")
+    public void testManuallySpecifiedIndices( final List<Path> bams, final List<Path> indices ) {
+        try ( final ReadsDataSource readsSource = new ReadsDataSource(bams, indices) ) {
+            Assert.assertTrue(readsSource.indicesAvailable(), "Explicitly-provided indices not detected for bams: " + bams);
+
+            final Iterator<GATKRead> queryReads = readsSource.query(new SimpleInterval("1", 1, 300));
+            int queryCount = 0;
+            while ( queryReads.hasNext() ) {
+                ++queryCount;
+                queryReads.next();
+            }
+            Assert.assertEquals(queryCount, 5, "Wrong number of reads returned in query");
+        }
+    }
+
+    @Test(dataProvider = "manuallySpecifiedIndexTestData")
+    public void testManuallySpecifiedIndicesNullIndexListOK( final List<Path> bams, final List<Path> indices ) {
+        try ( final ReadsDataSource readsSource = new ReadsDataSource(bams, (List<Path>)null) ) {
+            Assert.assertFalse(readsSource.indicesAvailable(), "Bams not indexed and explicit indices not provided, but indicesAvailable() returns true");
+        }
+    }
+
+    @Test(dataProvider = "manuallySpecifiedIndexTestData", expectedExceptions = UserException.class)
+    public void testManuallySpecifiedIndicesEmptyIndexList( final List<Path> bams, final List<Path> indices ) {
+        final ReadsDataSource readsSource = new ReadsDataSource(bams, Collections.emptyList());
+    }
+
+    @Test(dataProvider = "manuallySpecifiedIndexTestData", expectedExceptions = UserException.class)
+    public void testManuallySpecifiedIndicesWrongNumberOfIndices( final List<Path> bams, final List<Path> indices ) {
+        final List<Path> wrongIndices = new ArrayList<>();
+        wrongIndices.add(indices.get(0)); // Add one index, but not the other
+
+        final ReadsDataSource readsSource = new ReadsDataSource(bams, wrongIndices);
+    }
```

Comment:
Might be worth adding a test that uses a constructor that takes a index list and a custom SamReaderFactory, since thats the one that GATK always uses.